### PR TITLE
[Resource Graph] result format option added

### PR DESCRIPTION
**Pull request by OlhaTkachenko**
## Context:
Currently Resource Graph returns response in Table format:
data: { column: [],  rows: [] }
This format is good for portal, but is different from ARM response. 
Customers are doing custom conversion from that format to array format on their side. 
They are asking us to support array format also. For example, Adobe, VM blade were asking for it.
 
We are expanding result format with new option array for objects:
      data: []
 
## Problem:
Currently, the problem is how to achieve it with less overhead for customers.

We were considering options:
**1. New operation** (/Microsoft.ResourcesTopology/listresources **/array** ?api-version=2019-04-01)
PROS: no breaking change
CONS: confusing for customers, hard to add 3rd response type

**2. New property in query result**
If current format is requested:   {"data": {...},   "objects": null}
if new format:                           {"data": null,   "objects": [ ... ]}
PROS: not confusing
CONS: breaking change, require new api-version, hard to add 3rd response type, customers will wait longer for feature

**3. Return Table or Array from existing operation depending on provided value in query param ResultFormat.**
PROS: easy to expand with new response type,  least overheads, customers like it, existing SDK version continues to work.
CONS: new SDK version is required to be published.

As a team we feel option-3 is the best for our customers and with least overheads, and only con is we need to generate new SDK version. _**Existing SDK version continues to work.**_

**Points:**
We agree that it isn’t ideal to make customers modify code after sdk update, but we saw such behavior in many standard libraries.
All considered designs aren’t ideal anyway and have its disadvantages. 

## Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

**Review by chiragg4u — APPROVED** [review]
Thanks, Olha for this change. Many customers/partners are waiting to consume the JSON format and this enhancement is very beneficial.
For other reviewers. Please note that the backend service changes have already been deployed and running.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/stable/2019-04-01/resourcegraph.json [hunk 1]
```diff
@@ -255,7 +267,7 @@
         },
         "data": {
           "description": "Query output in tabular format.",
-          "$ref": "#/definitions/Table"
+          "type": "object"
```
Comment (praries880) [inline specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/stable/2019-04-01/resourcegraph.json]: why is the data type changing?

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/stable/2019-04-01/resourcegraph.json [hunk 2]
```diff
@@ -375,7 +387,7 @@
         },
         "data": {
           "description": "A table containing the desired facets. Only present if the facet is valid.",
-          "$ref": "#/definitions/Table"
+          "type": "object"
```
Comment (praries880) [inline specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/stable/2019-04-01/resourcegraph.json]: why is the type changing?

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/stable/2019-04-01/resourcegraph.json (hunk 1 above)
Comment (OlhaTkachenko) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/stable/2019-04-01/resourcegraph.json]: Before we were returning response in Table format, but now it is either Table or IList<Dictionary>. Please let us know if you know better way of doing it.

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/stable/2019-04-01/resourcegraph.json (hunk 2 above)
Comment (OlhaTkachenko) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/stable/2019-04-01/resourcegraph.json]: Same as above, we added new result format, which depends from query param ResultFormat

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/stable/2019-04-01/resourcegraph.json (hunk 1 above)
Comment (praries880) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/stable/2019-04-01/resourcegraph.json]: This looks like a breaking change

Comment (sanjaiganesh) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/stable/2019-04-01/resourcegraph.json]: yes. Changing response format is a breaking change. Please introduce new api version for these changes.
You could consider discriminator pattern that will allow ability to specify different properties based on "kind" property (top level property on arm resource).
I shall forward the details.. you may take a look


---
In reply to: [289210711](https://github.com/Azure/azure-rest-api-specs/pull/6148#discussion_r289210711) [](ancestors = 289210711)

**Comment by sanjaiganesh** [comment]
Adding ARM SignOff .. Based on offline discussion with ARG service team, It looks like a breaking change but it is actually not. the service ensures to detect the SDK calling in and do the right thing to continue return table format (based on resultFormat, which would null for old/current sdks).

File: specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/stable/2019-04-01/resourcegraph.json (hunk 1 above)
Comment (OlhaTkachenko) [reply specification/resourcegraph/resource-manager/Microsoft.ResourceGraph/stable/2019-04-01/resourcegraph.json]: Added description with more details and explanation, why it won't break but require sdk update.

**Review by praries880 — APPROVED** [review]

**Pull request merged**